ttnn/cpp/ttnn/operations/experimental/transformer/nlp_create_qkv_heads_decode/device/kernels/reader_tm_tile_layout_nlp_create_qkv_heads_decode.cpp

```diff
@@ -32,109 +33,129 @@ void kernel_main() {
     uint32_t qkv_y = 0;
     uint32_t total_input_cores = num_x * num_y;
     uint32_t num_tiles_per_core = head_size_num_tiles * (num_q_heads + 2 * num_kv_heads) / total_input_cores;
+    uint32_t num_q_cores = num_tiles_per_core * num_q_heads * head_size_num_tiles;
+    uint32_t num_kv_cores = num_tiles_per_core * num_kv_heads * head_size_num_tiles;
 
     uint64_t qkv_read_addr = get_noc_addr(in0_mcast_noc_x[qkv_x], in0_mcast_noc_y[qkv_y], q_start_addr) + in_tile_offset_by_batch;
     uint32_t num_tiles_read_cur_core = 0;
     uint32_t q_write_addr = 0;
     uint32_t tile_size = head_size/head_size_num_tiles;
 
-    for (uint32_t q = 0; q < num_q_heads; ++q) {
-        uint32_t wptr_offset = q < 16 ? q * SUBTILE_LINE_BYTES : (q - 16) * SUBTILE_LINE_BYTES + 512*ELEMENT_SIZE;
-        uint32_t q_write_addr = get_write_ptr(cb_id_q_out) + wptr_offset;
-        for (uint32_t i = 0; i < head_size_num_tiles; ++i) {
-            // Read first phase
-            if constexpr (PHASES_TO_READ == 0 || PHASES_TO_READ == 1) {
-                noc_async_read(qkv_read_addr, q_write_addr, SUBTILE_LINE_BYTES);
-                //noc_async_read_barrier();
-            }
-            // Read second phase
-            if constexpr (PHASES_TO_READ == 0 || PHASES_TO_READ == 2) {
-                noc_async_read(qkv_read_addr+256*ELEMENT_SIZE, q_write_addr+256*ELEMENT_SIZE, SUBTILE_LINE_BYTES);
-                //noc_async_read_barrier();
-            }
-
-            qkv_read_addr += tile_size;
-            q_write_addr += tile_size;
-            num_tiles_read_cur_core++;
+    // Skip Q section if PROCESS_QV is False
+    if constexpr (PROCESS_QV == 1) {
+        for (uint32_t q = 0; q < num_q_heads; ++q) {
+            uint32_t wptr_offset = q < 16 ? q * SUBTILE_LINE_BYTES : (q - 16) * SUBTILE_LINE_BYTES + 512*ELEMENT_SIZE;
```

Comment:
what are all those hardcoded numbers? can we either define constants with food names or comment what that means?

Reply:
I believe these hardcoded values are related to sub tiles and were there in the original form, so didn't care to remove them. But I can try parameterize them

Reply:
If you need to investigate what they mean, then nvm, but would be more readable for the next person if it were parameterized/documented what they mean.

Reply:
@caixunshiren these are for faces, right?

Reply:
Replaced them with parameters for readability